rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -1029,30 +1051,83 @@ object BuildNodeConfiguration extends Loggable {
                   merge(nodeInfo, agent.agentType, mode, seq.toList)
                 }
       }
-      //now change remaining BoundPolicyDraft to Policy, managing tracking variable values
-      drafts =  merged ++ (keptUniqueDraft ++ groupedDrafts.multiDirectives).map { d =>
-                  d.toPolicy(agent.agentType)
-                }
-      // check that agent is supported by all techniques
-      _      <- sequence(drafts) { d =>
-                  if(d.technique.agentConfigs.exists( _.agentType == agent.agentType)) {
-                    Full("ok")
-                  } else {
-                    Failure(s"Technique '${d.technique.id}' does not support agent '${agent.agentType}' which is needed for node '${nodeInfo.hostname}' (${nodeInfo.id.value}).")
-                  }
-                }
+      // upgrade other drafts to policy
+      others <-  { import cats.implicits._
+                   (keptUniqueDraft ++ groupedDrafts.multiDirectives).toList.traverse( _.toPolicy(agent.agentType) )
+                 }
     } yield {
       // we are sorting several things in that method, and I'm not sure we want to sort them in the same way (and so
       // factorize out sorting core) or not. We sort:
       // - policies based on non-multi-instance technique, by priority and then rule/directive order, to choose the correct one
       // - in a multi-instance, mono-policy case, we sort directives in the same fashion (but does priority make sense here?)
       // - and finally, here we sort all drafts (in that case, only by bundle order)
 
-      drafts.sortWith { case(d1, d2) =>
+      (merged ++ others).sortWith { case(d1, d2) =>
         BundleOrder.compareList(List(d1.ruleOrder, d1.directiveOrder), List(d2.ruleOrder, d2.directiveOrder)) <= 0
       }.toList
     }
   }
+
+  /*
+   * This method take all the hooks with the corresponding reportId / mode, and
+   * merge what should be merged.
+   * Order is not kept
+   */
+  def mergeRunHooks(policies: List[Policy], nodePolicyMode: Option[PolicyMode], globalPolicyMode: GlobalPolicyMode): List[NodeRunHook] = {
+    /*
+     * Hooks are merge:
+     *  - for the same kind, name and parameter (exactly the same, "parameters" order is critical)
+     * Then:
+     *  - append all condition (they will be "or", but the syntax is agent specific)
+     *  - append pair of (reportId, PolicyMode)
+     *
+     * If at least one of kind, name, parameter is not the same between two hooks, the are considered different.
+     *
+     * The merge must keep both the List[Policy] order AND in a Policy, the List[RunHook] order. So we
+     * avoid groupBy
+     */
+    // utility class to store reportId + RunHook
+    case class BoundHook(id: PolicyId, mode: PolicyMode, hook: RunHook)
+
+    def recMerge(currentHook: BoundHook, remaining: List[BoundHook]): List[NodeRunHook] = {
```

Comment:
this looks overly complicated for the task
couldn't you do a list of all BoundHook, and then group them by hook.kind; hook.name; hook.parameters  ?

Reply:
no, I need to keep the same order, groupBy use a map (a set actually), and it would loose the order (and it was harder to try to restor it after merge)

Reply:
you are totally right